File: AdventOfCode2024/ruby/lib/day3/day_3.rb
class Day3
    INPUT_PATH = File.expand_path('../../../../inputs/day3/input.txt', __FILE__).freeze
    MUL_PATTERN = /mul\((?<a>\d+),(?<b>\d+)\)/
    DO_PATTERN = /do\(\)/
    DONT_PATTERN = /don't\(\)/

    def self.puzzle1
        content = File.read(INPUT_PATH)
        content.scan(MUL_PATTERN).map { |a, b| parse_multiplication("mul(#{a},#{b})") }.sum
    end


    def self.puzzle2
        content = File.read(INPUT_PATH)
        process_with_instructions(content)
    end

    private

    def self.process_with_instructions(content)
        sum = 0
        enabled = true
        position = 0

        while position < content.length
            do_matcher = content[position..-1].match(DO_PATTERN)
            dont_matcher = content[position..-1].match(DONT_PATTERN)
            mul_matcher = content[position..-1].match(MUL_PATTERN)

            if do_matcher && do_matcher.begin(0) == 0
                enabled = true
                position += do_matcher.end(0)
            elsif dont_matcher && dont_matcher.begin(0) == 0
                enabled = false
                position += dont_matcher.end(0)
            elsif mul_matcher && mul_matcher.begin(0) == 0 && enabled
                sum += parse_multiplication(mul_matcher.to_s)
                position += mul_matcher.end(0)
            else
                position += 1
            end
        end

        sum
    end

    def self.parse_multiplication(expression)
      numbers = expression.gsub('mul(', '').gsub(')', '').split(',')
      begin
        num1, num2 = numbers.map(&:to_i)
        num1 * num2
      rescue
        "Invalid input"
      end
    end
end

puts Day3.puzzle1
puts Day3.puzzle2

File: AdventOfCode2024/java/src/main/java/day3/Day3.java
package day3;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;
import java.util.regex.Pattern;
public class Day3 {

    public static final String INPUT_PATH = "../../../../inputs/day3/input.txt";
    public static final Pattern MUL_PATTERN = Pattern.compile("mul\\((?<a>-?\\d+),(?<b>-?\\d+)\\)");
    public static final Pattern DO_PATTERN = Pattern.compile("do\\(\\)");
    public static final Pattern DONT_PATTERN = Pattern.compile("don't\\(\\)");


    public static void main(String[] args) throws FileNotFoundException {
        System.out.println(puzzle1());
        System.out.println(puzzle2());
    }

    public static int puzzle1() throws FileNotFoundException {
        try (Scanner scanner = new Scanner(new File(INPUT_PATH))) {
            return scanner.findAll(MUL_PATTERN)
                    .map(match -> match.group().toString())
                    .mapToInt(Day3::parseMultiplication)
                    .sum();
        }
    }

    public static int puzzle2() throws FileNotFoundException {
        try (Scanner scanner = new Scanner(new File(INPUT_PATH))) {
            String context = scanner.useDelimiter("\\Z").next();
            return processWithInstructions(context);
        }
    }

    private static int processWithInstructions(String context) {
        int sum = 0;
        boolean enabled = true;
        int position = 0;

        while(position < context.length()) {
            var doMatcher = DO_PATTERN.matcher(context.substring(position));
            var dontMatcher = DONT_PATTERN.matcher(context.substring(position));
            var mulMatcher = MUL_PATTERN.matcher(context.substring(position));

            if (doMatcher.find() && doMatcher.start() == 0) {
                enabled = true;
                position += doMatcher.end();
            } else if (dontMatcher.find() && dontMatcher.start() == 0) {
                enabled = false;
                position += dontMatcher.end();
            } else if (mulMatcher.find() && mulMatcher.start() == 0) {
                if (enabled) {
                    sum += parseMultiplication(mulMatcher.group().toString());
                }
                position += mulMatcher.end();
            } else {
                position++;
            }
        }

        return sum;
    }


    private static int parseMultiplication(String expression) {
        try {
            String[] numbers = expression.replace("mul(", "").replace(")", "").split(",");
            int num1 = Integer.parseInt(numbers[0]);
            int num2 = Integer.parseInt(numbers[1]);
            return num1 * num2;
        } catch (Exception e) {
            return 0;
        }
    }
}
